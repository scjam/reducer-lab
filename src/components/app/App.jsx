/* eslint-disable max-len */
import React, { useReducer } from 'react';
import useColor, { initialState } from '../../reducers/useRecord';

// const useRecord = (init) => {
//   const [before, setBefore] = useState([]);
//   const [current, setCurrent] = useState(init);
//   const [after, setAfter] = useState([]);

//   const undo = () => {
//     setAfter(after => [current, ...after]);
//     setCurrent(before[before.length - 1]);
//     setBefore(before => before.slice(0, -1));
//   };

//   const redo = () => {
//     setBefore(before => [...before, current]);
//     setCurrent(after[0]); 
//     setAfter(after => after.slice(1));
//   };

//   const record = val => {
//     setBefore(before => [...before, current]);
//     setCurrent(val);
//   };

//   return {
//     undo,
//     record,
//     redo,
//     current,
//   };
// };

function App() {
  // const { current, undo, redo, record } = useRecord('#FF0000');
  const [state, dispatch] = useReducer(useColor, initialState);

  const undo = () => {
    dispatch({ type: 'UNDO' });
  };

  const redo = () => {
    dispatch({ type: 'REDO' });
  };

  return (
    <>
      <button data-testid="undo" onClick={undo} disabled={!state.before.length} >undo</button>
      <button data-testid="redo" onClick={redo} disabled={!state.after.length}>redo</button>
      <label htmlFor="RECORD">color input</label>
      <input 
        id="RECORD" 
        type="color" 
        value={state.current} 
        onChange={({ target }) => dispatch({
          payload: target.value,
          type: target.id })} 
      />
      <div data-testid="colorDiv" style={{ backgroundColor: state.current, width: '10rem', height: '10rem' }}></div>
    </>
  );
}

export default App;
