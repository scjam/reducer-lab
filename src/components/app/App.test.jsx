import React from 'react';
import { render, screen, cleanup, fireEvent } from '@testing-library/react';
import App from './App';

describe('colorpicker change, undo, redo, and record', () => {
  afterEach(() => cleanup());
  it('renders App', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('changes background color', async() => {
    render(<App />);
    const colorInput = screen.getByLabelText('color input');
    const div = await screen.getByTestId('colorDiv');
    fireEvent.change(colorInput, {
      target: {
        value: '#00FF00'
      }
    });
    expect(div).toHaveStyle({
      backgroundColor: '#00FF00'
    });
  });

  it('should undo previous color change', async() => {
    render(<App />);
    const colorInput = screen.getByLabelText('color input');
    const undo = screen.getByTestId('undo');
    fireEvent.change(colorInput, {
      target: {
        value: '#0000FF'
      }
    });
    fireEvent.click(undo);
    const div = await screen.findByTestId('colorDiv');
    expect(div).toHaveStyle({
      backgroundColor: '#000000'
    });
  });

  it('should redo the previous color change', async() => {
    render(<App />);
    const colorInput = screen.getByLabelText('color input');
    const undo = screen.getByTestId('undo');
    const redo = screen.getByTestId('redo');
    fireEvent.change(colorInput, {
      target: {
        value: '#0000FF'
      }
    });
    fireEvent.click(undo);
    fireEvent.click(redo);
    const div = await screen.getByTestId('colorDiv');
    expect(div).toHaveStyle({
      backgroundColor: '#0000FF'
    });    
  });
  
  
});
