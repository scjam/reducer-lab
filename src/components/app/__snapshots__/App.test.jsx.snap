// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`colorpicker change, undo, redo, and record renders App 1`] = `
<DocumentFragment>
  <button
    data-testid="undo"
    disabled=""
  >
    undo
  </button>
  <button
    data-testid="redo"
    disabled=""
  >
    redo
  </button>
  <label
    for="RECORD"
  >
    color input
  </label>
  <input
    id="RECORD"
    type="color"
    value="#000000"
  />
  <div
    data-testid="colorDiv"
    style="background-color: rgb(0, 0, 0); width: 10rem; height: 10rem;"
  />
</DocumentFragment>
`;
